frame/vesting/src/lib.rs

```diff
@@ -583,7 +914,87 @@ mod tests {
 				assert_eq!(Vesting::vesting_balance(&2), Some(0)); // Account 2 has fully vested by block 30
 				assert_eq!(Vesting::vesting_balance(&12), Some(0)); // Account 2 has fully vested by block 30
+			});
+	}
+
+	#[test]
+	fn check_vesting_status_for_multi_schedule_account() {
+		ExtBuilder::default().existential_deposit(ED).build().execute_with(|| {
+			assert_eq!(System::block_number(), 1);
+			let sched0 = VestingInfo::try_new::<Test>(
+				ED * 20,
+				ED, // Vesting over 20 blocks
+				10,
+			)
+			.unwrap();
+
+			// Use 2 already has a vesting schedule.
+			assert_eq!(Vesting::vesting(&2).unwrap(), vec![sched0]);
+
+			// User 2's free balance is from sched0
+			let free_balance = Balances::free_balance(&2);
+			assert_eq!(free_balance, ED * (20));
+			assert_eq!(Vesting::vesting_balance(&2), Some(free_balance));
+
+			// Add a 2nd schedule that is already unlocking by block #1
+			let sched1 = VestingInfo::try_new::<Test>(
+				ED * 10,
+				ED, // Vesting over 10 blocks
+				0,
+			)
+			.unwrap();
+			assert_ok!(Vesting::vested_transfer(Some(4).into(), 2, sched1));
+			// Free balance is equal to the two existing schedules total amount.
+			let free_balance = Balances::free_balance(&2);
+			assert_eq!(free_balance, ED * (10 + 20));
+			// The most recently added schedule exists.
+			assert_eq!(Vesting::vesting(&2).unwrap(), vec![sched0, sched1]);
+			// sched1 has free funds at block #1, but nothing else.
+			assert_eq!(Vesting::vesting_balance(&2), Some(free_balance - sched1.per_block()));
+
+			// Add a 3rd schedule
+			let sched2 = VestingInfo::try_new::<Test>(
+				ED * 30,
+				ED, // Vesting over 30 blocks
+				5,
+			)
+			.unwrap();
+			assert_ok!(Vesting::vested_transfer(Some(4).into(), 2, sched2));
+
+			System::set_block_number(9);
+			// Free balance is equal to the 3 existing schedules total amount.
+			let free_balance = Balances::free_balance(&2);
+			assert_eq!(free_balance, ED * (10 + 20 + 30));
+			// sched1 and sched2 are freeing funds at block #9.
+			assert_eq!(
+				Vesting::vesting_balance(&2),
+				Some(free_balance - sched1.per_block() * 9 - sched2.per_block() * 4)
+			);
+
+			System::set_block_number(20);
+			// At block #20 sched1 is fully unlocked while sched2 and sched0 are partially unlocked.
+			assert_eq!(
+					Vesting::vesting_balance(&2),
+					Some(
+						free_balance -
+							sched1.locked() - sched2.per_block() * 15 -
+							sched0.per_block() * 10
+					)
+				);
 
+				System::set_block_number(30);
+				// At block #30 sched0 and sched1 are fully unlocked while sched2 is partially unlocked.
+				assert_eq!(
+					Vesting::vesting_balance(&2),
+					Some(
+						free_balance - sched1.locked() - sched2.per_block() * 25 - sched0.locked()
+					)
+				);
+
+				// At block #35 sched2 fully unlocks and thus all schedules funds are unlocked.
+				System::set_block_number(35);
+				assert_eq!(System::block_number(), 35);
+				assert_eq!(Vesting::vesting_balance(&2), Some(0));
 			});
 	}
 
```

Comment:
* shows single vesting is working just like before
* shows multiple vesting works as expected
* merges two ongoing vesting schedules
* merges one ended and one ongoing schedule
* merges to ended schdules
* all dispatcherror paths are checked as noop <-- (a lot, and its important)

Reply:
hoisted this to the top comment